scipy/sparse/compressed.py

```diff
@@ -713,10 +713,11 @@ def _major_slice(self, idx, copy=False):
         if M == 0:
             return self.__class__(new_shape)
 
-        row_nnz = np.diff(self.indptr)
+        indices = np.arange(start, stop, step)
+        row_nnz = self.indptr[indices + 1] - self.indptr[indices]
```

Comment:
This I think doesn't handle out-of-bounds slice indices, which are generally allowed in Python (do we have a test for that?).

Maybe, can be written as `self.indptr[start2:stop2:step] - self.indptr[start:stop:step]`, with start2=start2+1, stop2=stop2+1 but taking care of the case when the start/stop indices are `-1`

Reply:
I don't think the start/stop indices can be `-1` can they? They are resolved to true indices by `idx.indices(M)`?

Reply:
And it should be out-of-bounds safe I think since the indexes given by `np.arange(start, stop, step)` should be in-bounds for `self` and my understanding is that `self.indptr.shape[0] == self.shape[0] + 1`?

Reply:
Similarly:
```python
>>> slice(0, 22, None).indices(5)
(0, 5, 1)
```

Reply:
Ok, yes, that's right. You might be able to avoid constructing the arange however.

Reply:
If I do `row_nnz = self.indptr[(start + 1):(stop + 1):step] - self.indptr[start:stop:step]` I get errors to do with the sizes being different.

```
FAILED scipy/sparse/tests/test_base.py::TestCSR::test_slicing_3 - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::TestCSC::test_slicing_3 - ValueError: operands could not be broadcast together with shapes (4,) (5,)
FAILED scipy/sparse/tests/test_base.py::TestCSC::test_fancy_indexing - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::TestCSRNonCanonical::test_slicing_3 - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::TestCSCNonCanonical::test_slicing_3 - ValueError: operands could not be broadcast together with shapes (4,) (5,)
FAILED scipy/sparse/tests/test_base.py::TestCSCNonCanonical::test_fancy_indexing - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_limit_10[TestCSC-test_fancy_indexing] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_limit_10[TestCSC-test_slicing_3] - ValueError: operands could not be broadcast together with shapes (4,) (5,)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_limit_10[TestCSR-test_slicing_3] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_random[TestCSC-test_fancy_indexing] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_random[TestCSC-test_slicing_3] - ValueError: operands could not be broadcast together with shapes (4,) (5,)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_random[TestCSR-test_slicing_3] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_32[TestCSC-test_fancy_indexing] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_32[TestCSC-test_slicing_3] - ValueError: operands could not be broadcast together with shapes (4,) (5,)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_32[TestCSR-test_slicing_3] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_64[TestCSC-test_fancy_indexing] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_64[TestCSC-test_slicing_3] - ValueError: operands could not be broadcast together with shapes (4,) (5,)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_resiliency_all_64[TestCSR-test_slicing_3] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_no_64[TestCSC-test_fancy_indexing] - ValueError: provided out is the wrong size for the reduction
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_no_64[TestCSC-test_slicing_3] - ValueError: operands could not be broadcast together with shapes (4,) (5,)
FAILED scipy/sparse/tests/test_base.py::Test64Bit::test_no_64[TestCSR-test_slicing_3] - ValueError: provided out is the wrong size for the reduction
```

I do think that generating the set of indices first is the safest method. There's an open question as to whether that would be better achieved with `np.arange` or just `range`?

Reply:
Test:
```python
import timeit
import numpy as np


X = np.random.randint(0, 100, 1000)
start = 2
stop = 53
step = 4


print("range:", timeit.timeit(lambda: X[range(start, stop, step)], number=1000000))
print("np.arange:", timeit.timeit(lambda: X[np.arange(start, stop, step)], number=1000000))
```

Result:
```python
range: 2.285575583
np.arange: 0.42806620900000025
```

So looks to me like `np.arange` is the right tool for the job?

Reply:
I guess the problem is that `slice(4, None, -2).indices(10) == (4, -1, -2)` so stop can be -1 if step is negative, which would need separate handling.

Reply:
Worked it out!

Reply:
Test:
```python
import timeit
import numpy as np


X = np.random.randint(0, 100, 1000)
start = 4
stop = None
step = -2


def slicing():
    start2 = None if start == -1 else start
    stop2 = None if stop == -1 else stop
    start3 = None if start2 is None else start2 + 1
    stop3 = None if stop2 is None else stop2 + 1
    X[start3:stop3:step] - X[start2:stop2:step]

def _numpy():
    indices = np.arange(start, stop, step)
    X[indices + 1] - X[indices]

print("slicing:", timeit.timeit(slicing, number=1000000))
print("np.arange:", timeit.timeit(_numpy, number=1000000))
```

Result:
```python
slicing: 0.51986475
np.arange: 1.110926584
```

Reply:
Added a few tests that shook out some problems (slice(None, None, -1), slice(0, None, -1), etc), plus another suggestion for this, it's a bit fiddly. I guess now it works, but needs still another thinking through the logic.

Reply:
```python
import timeit
import numpy as np


X = np.random.randint(0, 100, 1000)
idx = slice(4, 10, 2)
start, stop, step = idx.indices(X.shape[0] -1)

def without_slicing():
    stop2 = stop
    if stop == -1 and start >= 0:
        stop2 = None
    return X[(start + 1):(stop + 1):step] - X[start:stop2:step]

def with_slicing():
    if stop == -1 and start >= 0:
        idx0 = slice(start, None, step)
    else:
        idx0 = slice(start, stop, step)
    idx1 = slice(start + 1, stop + 1, step)
    return X[idx1] - X[idx0]

print("with_slicing:", timeit.timeit(with_slicing, number=1000000))
print("without_slicing:", timeit.timeit(without_slicing, number=1000000))
```

```python
with_slicing: 0.583187417
without_slicing: 0.45088366700000004
```

Seems like there's a bit of unnecessary overhead in initialising the `slice` objects. Maybe better to stick with `[start:stop:step]` notation?